File: go/internal/domains/agent/tools.go
package agent

// The tool belt: a registry that NEVER crashes (missing tool / bad input -> a graceful error observation) and
// the built-ins — calc (a recursive-descent parser over + - * / % and parentheses; NEVER an eval; one-decimal
// output so all three languages produce identical observations) and echo.

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

func runTool(name string, args map[string]string) (string, bool) {
	switch name {
	case "calc":
		return calc(args["expr"]), true
	case "echo":
		return args["text"], true
	default:
		return fmt.Sprintf("tool '%s' not found", name), false
	}
}

// ── calc: tokenizer + recursive descent (expr := term (('+'|'-') term)* · term := unary (('*'|'/'|'%') unary)*
//    · unary := '-' unary | atom · atom := number | '(' expr ')') ──────────────────────────────────────────────

type calcParser struct {
	tokens []string
	pos    int
}

func calc(expr string) string {
	p := &calcParser{tokens: calcTokens(expr)}
	if len(p.tokens) == 0 {
		return "error: invalid expression"
	}
	v, ok := p.parseExpr()
	if !ok || p.pos != len(p.tokens) {
		return "error: invalid expression"
	}
	if math.IsInf(v, 0) || math.IsNaN(v) { // ±Inf (1/0, overflow) / NaN -> invalid, matching python/node (one observation ×3)
		return "error: invalid expression"
	}
	return fmt.Sprintf("%.1f", v)
}

func calcTokens(expr string) []string {
	out := []string{}
	i := 0
	for i < len(expr) {
		c := expr[i]
		switch {
		case c == ' ':
			i++
		case strings.ContainsRune("+-*/%()", rune(c)):
			out = append(out, string(c))
			i++
		case (c >= '0' && c <= '9') || c == '.':
			j := i
			for j < len(expr) && ((expr[j] >= '0' && expr[j] <= '9') || expr[j] == '.') {
				j++
			}
			out = append(out, expr[i:j])
			i = j
		default:
			return nil // any other character -> invalid (never executed, never evaluated)
		}
	}
	return out
}

func (p *calcParser) peek() string {
	if p.pos < len(p.tokens) {
		return p.tokens[p.pos]
	}
	return ""
}

func (p *calcParser) parseExpr() (float64, bool) {
	v, ok := p.parseTerm()
	if !ok {
		return 0, false
	}
	for p.peek() == "+" || p.peek() == "-" {
		op := p.tokens[p.pos]
		p.pos++
		r, ok := p.parseTerm()
		if !ok {
			return 0, false
		}
		if op == "+" {
			v += r
		} else {
			v -= r
		}
	}
	return v, true
}

func (p *calcParser) parseTerm() (float64, bool) {
	v, ok := p.parseUnary()
	if !ok {
		return 0, false
	}
	for p.peek() == "*" || p.peek() == "/" || p.peek() == "%" {
		op := p.tokens[p.pos]
		p.pos++
		r, ok := p.parseUnary()
		if !ok {
			return 0, false
		}
		switch op {
		case "*":
			v *= r
		case "/":
			v /= r
		default:
			if int64(r) == 0 { // int64 % 0 PANICS in Go (uncontained crash) -> reject, matching python/node 1%0
				return 0, false
			}
			v = float64(int64(v) % int64(r))
		}
	}
	return v, true
}

func (p *calcParser) parseUnary() (float64, bool) {
	if p.peek() == "-" {
		p.pos++
		v, ok := p.parseUnary()
		return -v, ok
	}
	if p.peek() == "(" {
		p.pos++
		v, ok := p.parseExpr()
		if !ok || p.peek() != ")" {
			return 0, false
		}
		p.pos++
		return v, true
	}
	tok := p.peek()
	if tok == "" {
		return 0, false
	}
	v, err := strconv.ParseFloat(tok, 64)
	if err != nil {
		return 0, false
	}
	p.pos++
	return v, true
}

File: node/src/domains/agent/tools.js
// The tool belt: a dispatcher that NEVER crashes (missing tool -> a graceful error observation) and the
// built-ins — calc (a recursive-descent parser over + - * / % and parentheses; NEVER an eval; one-decimal
// output so all three languages produce identical observations) and echo.

export function runTool(name, args) {
  if (name === 'calc') return { output: calc(String((args && args.expr) || '')), ok: true };
  if (name === 'echo') return { output: String((args && args.text) || ''), ok: true };
  return { output: `tool '${name}' not found`, ok: false };
}

function tokens(expr) {
  const out = [];
  let i = 0;
  while (i < expr.length) {
    const c = expr[i];
    if (c === ' ') { i++; continue; }
    if ('+-*/%()'.includes(c)) { out.push(c); i++; continue; }
    if ((c >= '0' && c <= '9') || c === '.') {
      let j = i;
      while (j < expr.length && ((expr[j] >= '0' && expr[j] <= '9') || expr[j] === '.')) j++;
      out.push(expr.slice(i, j));
      i = j;
      continue;
    }
    return null; // any other character -> invalid (never executed, never evaluated)
  }
  return out;
}

function calc(expr) {
  const toks = tokens(expr);
  if (!toks || toks.length === 0) return 'error: invalid expression';
  const p = { toks, pos: 0 };
  const v = parseExpr(p);
  if (v === null || p.pos !== p.toks.length) return 'error: invalid expression';
  if (!Number.isFinite(v)) return 'error: invalid expression'; // ±Infinity (1/0) / NaN (1%0) -> invalid, matching python/go (one observation ×3)
  return v.toFixed(1);
}

const peek = (p) => (p.pos < p.toks.length ? p.toks[p.pos] : '');

function parseExpr(p) {
  let v = parseTerm(p);
  if (v === null) return null;
  while (peek(p) === '+' || peek(p) === '-') {
    const op = p.toks[p.pos++];
    const r = parseTerm(p);
    if (r === null) return null;
    v = op === '+' ? v + r : v - r;
  }
  return v;
}

function parseTerm(p) {
  let v = parseUnary(p);
  if (v === null) return null;
  while (peek(p) === '*' || peek(p) === '/' || peek(p) === '%') {
    const op = p.toks[p.pos++];
    const r = parseUnary(p);
    if (r === null) return null;
    if (op === '*') v *= r;
    else if (op === '/') v /= r;
    else v = Math.trunc(v) % Math.trunc(r);
  }
  return v;
}

function parseUnary(p) {
  if (peek(p) === '-') {
    p.pos++;
    const v = parseUnary(p);
    return v === null ? null : -v;
  }
  if (peek(p) === '(') {
    p.pos++;
    const v = parseExpr(p);
    if (v === null || peek(p) !== ')') return null;
    p.pos++;
    return v;
  }
  const tok = peek(p);
  if (tok === '' || !/^[0-9.]+$/.test(tok)) return null;
  const v = Number(tok);
  if (!Number.isFinite(v)) return null;
  p.pos++;
  return v;
}
